op-node/sources/l2_client.go

```diff
@@ -0,0 +1,124 @@
+package sources
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-node/client"
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+	"github.com/ethereum-optimism/optimism/op-node/sources/caching"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type L2ClientConfig struct {
+	EthClientConfig
+
+	L2BlockRefsCacheSize int
+
+	Genesis rollup.Genesis
+}
+
+func L2ClientDefaultConfig(config *rollup.Config, trustRPC bool) *L2ClientConfig {
+	// Cache 3/2 worth of sequencing window of payloads, block references, receipts and txs
+	span := int(config.SeqWindowSize) * 3 / 2
+	// Estimate number of L2 blocks in this span of L1 blocks
+	// (there's always one L2 block per L1 block, L1 is thus the minimum, even if block time is very high)
+	if config.BlockTime < 12 && config.BlockTime > 0 {
+		span *= 12
+		span /= int(config.BlockTime)
+	}
+	if span > 1000 { // sanity cap. If a large sequencing window is configured, do not make the cache too large
+		span = 1000
+	}
+	return &L2ClientConfig{
+		EthClientConfig: EthClientConfig{
+			// receipts and transactions are cached per block
+			ReceiptsCacheSize:     span,
+			TransactionsCacheSize: span,
+			HeadersCacheSize:      span,
+			PayloadsCacheSize:     span,
+			MaxRequestsPerBatch:   20, // TODO: tune batch param
+			MaxConcurrentRequests: 10,
+			TrustRPC:              trustRPC,
+			MustBePostMerge:       true,
```

Comment:
I think that `MustBePostMerge` should be configurable, its only going to cause a headache for projects that want to use our codebase on altchains

Reply:
Ideally any sort of local validation should be configurable with defaults to true so that altcoins can explicitly opt out of security to run on their chain

Reply:
It's always `true` for L2, since the L2 is the post-merge chain we have with op-geth. For L1 it defaults to`false`, and although we could set it to `true` post-merge on ethereum, we don't strictly have to, so I'm going to leave the default as-is and revisit it later.